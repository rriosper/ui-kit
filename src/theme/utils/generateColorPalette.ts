import { darken, lighten, rgba } from 'polished'
import { mergeDeepLeft } from 'ramda'

function alpha (intensity, color) {
  return rgba(color, intensity)
}

const PALETTE_MODIFIERS = [
  {
    iterations: 9,
    key: 'A',
    fn: alpha
  },
  {
    iterations: 6,
    key: 'D',
    fn: darken
  },
  {
    iterations: 6,
    key: 'L',
    fn: lighten
  }
]

const arrayOf = (n: number): number[] => {
  const arr: number[] = []
  for (let i = 0; i < n; i++) {
    arr.push(i + 1)
  }
  return arr
}

const generateVariation = ({ fn, key, iterations }, color) => {
  const arr: number[] = arrayOf(iterations)
  return arr.reduce((acc, intensity) => {
    const procesedKey = `${key}${intensity + 1}`
    const processedIntensity = Number(`0.${intensity + 1}`)
    acc[procesedKey] = fn(processedIntensity, color)
    return acc
  }, {})
}

const generateColorPalette = (color) => {
  if (!color) {
    throw new Error('It is neccesary indicate a color')
  }

  const palette = PALETTE_MODIFIERS.reduce(
    (acc, modifier) => mergeDeepLeft(generateVariation(modifier, color), acc),
    { root: color }
  )

  return palette
}

export default generateColorPalette
