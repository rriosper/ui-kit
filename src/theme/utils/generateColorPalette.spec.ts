import generateColorPalette from './generateColorPalette'

test('generateColorPalette', () => {
  expect(generateColorPalette('red')).toMatchSnapshot()
  expect(generateColorPalette('blue')).toMatchSnapshot()
  expect(generateColorPalette('#333')).toMatchSnapshot()
})
