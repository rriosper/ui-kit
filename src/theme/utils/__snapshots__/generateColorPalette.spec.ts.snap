// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateColorPalette 1`] = `
Object {
  "A10": "rgba(255,0,0,0.1)",
  "A2": "rgba(255,0,0,0.2)",
  "A3": "rgba(255,0,0,0.3)",
  "A4": "rgba(255,0,0,0.4)",
  "A5": "rgba(255,0,0,0.5)",
  "A6": "rgba(255,0,0,0.6)",
  "A7": "rgba(255,0,0,0.7)",
  "A8": "rgba(255,0,0,0.8)",
  "A9": "rgba(255,0,0,0.9)",
  "D2": "#900",
  "D3": "#600",
  "D4": "#300",
  "D5": "#000",
  "D6": "#000",
  "D7": "#000",
  "L2": "#f66",
  "L3": "#f99",
  "L4": "#fcc",
  "L5": "#fff",
  "L6": "#fff",
  "L7": "#fff",
  "root": "red",
}
`;

exports[`generateColorPalette 2`] = `
Object {
  "A10": "rgba(0,0,255,0.1)",
  "A2": "rgba(0,0,255,0.2)",
  "A3": "rgba(0,0,255,0.3)",
  "A4": "rgba(0,0,255,0.4)",
  "A5": "rgba(0,0,255,0.5)",
  "A6": "rgba(0,0,255,0.6)",
  "A7": "rgba(0,0,255,0.7)",
  "A8": "rgba(0,0,255,0.8)",
  "A9": "rgba(0,0,255,0.9)",
  "D2": "#009",
  "D3": "#006",
  "D4": "#003",
  "D5": "#000",
  "D6": "#000",
  "D7": "#000",
  "L2": "#66f",
  "L3": "#99f",
  "L4": "#ccf",
  "L5": "#fff",
  "L6": "#fff",
  "L7": "#fff",
  "root": "blue",
}
`;

exports[`generateColorPalette 3`] = `
Object {
  "A10": "rgba(51,51,51,0.1)",
  "A2": "rgba(51,51,51,0.2)",
  "A3": "rgba(51,51,51,0.3)",
  "A4": "rgba(51,51,51,0.4)",
  "A5": "rgba(51,51,51,0.5)",
  "A6": "rgba(51,51,51,0.6)",
  "A7": "rgba(51,51,51,0.7)",
  "A8": "rgba(51,51,51,0.8)",
  "A9": "rgba(51,51,51,0.9)",
  "D2": "#000",
  "D3": "#000",
  "D4": "#000",
  "D5": "#000",
  "D6": "#000",
  "D7": "#000",
  "L2": "#666",
  "L3": "#808080",
  "L4": "#999",
  "L5": "#b3b3b3",
  "L6": "#ccc",
  "L7": "#e5e5e5",
  "root": "#333",
}
`;
